quicknes/bizinterface.cpp

```diff
@@ -224,118 +185,154 @@ EXPORT int qn_get_memory_area(Nes_Emu *e, int which, const void **data, int *siz
 		*data = e->high_mem();
 		*size = e->high_mem_size;
 		*writable = 1;
-		*name = "WRAM";
+		*name = "SRAM";
 		return 1;
 	case 2:
 		*data = e->chr_mem();
 		*size = e->chr_size();
-		*writable = 0;
+		*writable = 1;
 		*name = "CHR";
 		return 1;
 	case 3:
 		*data = e->nametable_mem();
 		*size = e->nametable_size();
-		*writable = 0;
+		*writable = 1;
 		*name = "CIRAM (nametables)";
 		return 1;
 	case 4:
 		*data = e->cart()->prg();
 		*size = e->cart()->prg_size();
-		*writable = 0;
+		*writable = 1;
 		*name = "PRG ROM";
 		return 1;
 	case 5:
 		*data = e->cart()->chr();
 		*size = e->cart()->chr_size();
-		*writable = 0;
+		*writable = 1;
 		*name = "CHR VROM";
 		return 1;
 	case 6:
 		*data = e->pal_mem();
-		*size = 32;
+		*size = e->pal_mem_size();
 		*writable = 1;
-		*name = "PALRAM";
+		*name = "Palette RAM";
 		return 1;
 	case 7:
-		*data = e->oam_mem();
-		*size = 256;
+	    *data = e->spr_mem();
+		*size = e->spr_mem_size();
 		*writable = 1;
-		*name = "OAM";
+		*name = "Sprite RAM";
 		return 1;
 	}
 }
 
-EXPORT unsigned char qn_peek_prgbus(Nes_Emu *e, int addr)
+EXPORT unsigned char qn_peek_prgbus(quickerNES::Emu *e, int addr)
 {
 	return e->peek_prg(addr & 0xffff);
 }
 
-EXPORT void qn_poke_prgbus(Nes_Emu *e, int addr, unsigned char val)
+EXPORT void qn_poke_prgbus(quickerNES::Emu *e, int addr, unsigned char val)
 {
 	e->poke_prg(addr & 0xffff, val);
 }
 
-EXPORT void qn_get_cpuregs(Nes_Emu *e, unsigned int *dest)
+EXPORT void qn_get_cpuregs(quickerNES::Emu *e, unsigned int *dest)
 {
 	e->get_regs(dest);
 }
 
-EXPORT const char *qn_get_mapper(Nes_Emu *e, int *number)
+EXPORT const char *qn_get_mapper(quickerNES::Emu *e, int *number)
 {
 	int m = e->cart()->mapper_code();
 	if (number)
 		*number = m;
 	switch (m)
 	{
 	default: return "unknown";
-	case 0: return "nrom";
-	case 1: return "mmc1";
-	case 2: return "unrom";
-	case 3: return "cnrom";
-	case 4: return "mmc3";
-	case 7: return "aorom";
-	case 69: return "fme7";
-	case 5: return "mmc5";
-	case 19: return "namco106";
-	case 24: return "vrc6a";
-	case 26: return "vrc6b";
-	case 11: return "color_dreams";
-	case 34: return "nina1";
-	case 66: return "gnrom";
-	case 87: return "mapper_87";
+	case   0: return "nrom";
+	case   1: return "mmc1";
+	case   2: return "unrom";
+	case   3: return "cnrom";
+	case   4: return "mmc3";
+	case   5: return "mmc5";

```

Comment:
How complete is the MMC5 support?  Can it run all commercial MMC5 games with no glitches?

I ask because the original QuickNES claimed MMC5 support but it was very buggy.

Reply:
I just tried running `Castlevania III - Dracula's Curse (U) [!].nes` and it outright failed. I'll have to investigate further; perhaps it's something I can fix in QuickerNES

Reply:
Update: I tested the game using my own frontend to QuickerNES and it seems to emulate just fine (no visible glitches, although sound needs to be checked). The crash I describe above seems to happen in the interface between BizHawk and the emulation core. I'll go ahead and get it fixed

Reply:
I just traced the error to the implementation of the serialization function in the original QuickNES. Turns out games like Castlevania III cannot tell you how much size their states will need until they have ran for a few frames (until they determine the real size of their nametable size). The solution is to assume always the largest possible NTAB size. I'm implementing the fix now and adding an automated test in QuickerNES to make sure it doesn't happen again

Reply:
Ok, issue fixed. Castlevania 3 seems to be running fine now. Here's a little demo:

https://youtu.be/icrWn6yXnjo